zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/flexible/QuorumOracleMaj.java

```diff
@@ -0,0 +1,158 @@
+        super(props);
+        needOracle  = new AtomicBoolean(false);
+        oracle = null;
+        setOracle(oraclePath);
+    }
+
+    private void setOracle(String path) {
+        if (oracle == null) {
+            oracle = path;
+            LOG.info("Oracle is set to {}", path);
+        } else {
+            LOG.warn("Oracle is already set. Ignore:{}", path);
+        }
+    }
+
+    @Override
+    public void updateNeedOracle(List<LearnerHandler> forwardingFollowers) {
+        // We do not have any follower, and our voting member is 2
+        needOracle.set(forwardingFollowers.isEmpty() && super.getVotingMembers().size() == 2);
+    }
+
+    @Override
+    public boolean askOracle() {
+
+        FileReader fr = null;
+        try {
+            int read;
+            fr = new FileReader(oracle);
+            read = fr.read();
+            LOG.debug("Oracle says:{}", (char) read);
+            return (char) read == '1';
+        } catch (Exception e) {
+            if (oracle == null) {
+                e.printStackTrace();
+                LOG.error("Oracle is not set, return false");
+            }
+            return false;
+        } finally {
+            if (fr != null) {
+                try {
+                    fr.close();
+                } catch (IOException e) {
+                    LOG.debug("Fail to close inputStream");
+                }
+            }
+        }
+    }
+
+    @Override
+    public boolean getNeedOracle() {
+        return needOracle.get() && super.getVotingMembers().size() == 2;
+    }
+
+    @Override
+    public String getOraclePath() {
+        return oracle;
+    }
+
+    @Override
+    public boolean containsQuorum(Set<Long> ackSet) {
+        if (oracle == null || getVotingMembers().size() > 2) {
+            return super.containsQuorum(ackSet);
+        } else if (!super.containsQuorum(ackSet)) {
+            if (getNeedOracle()) {
+                LOG.debug("We lose the quorum, but we do not have any valid followers Oracle:{}", askOracle());
+                return askOracle();
+            } else {
+                return false;
+            }
+        } else {
+            return true;
+        }
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        QuorumOracleMaj qm = (QuorumOracleMaj) o;
+        if (qm.getVersion() == super.getVersion()) {
+            return true;
+        }
+        if (super.getAllMembers().size() != qm.getAllMembers().size()) {
+            return false;
+        }
+        for (QuorumPeer.QuorumServer qs : super.getAllMembers().values()) {
+            QuorumPeer.QuorumServer qso = qm.getAllMembers().get(qs.id);
+            if (qso == null || !qs.equals(qso)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public int hashCode() {
```

Comment:
why don't we just use the default implementation if it is arbitrary?

Reply:
I have the same question as well. I made this because QuorumMaj, QuorumOracleMaj's superclass, overrides hasCode() and equals(), the findbug asks me to override them.